tests/integration/gregor/test_caf_generation.py

```diff
@@ -0,0 +1,217 @@
+        print("ValueError: has GCS_OAUTH_TOKEN expired?", e)
+        raise e
+
+
+def test_allele_counts_first_5_rows(chr3_vcf_path, phenotype_table):
+    """test that the calculated allele counts with no phenotype specified matches
+    the actual counts stored in the INFO of the first 10 rows. Works for diploid (non-sex) variants
+    """
+
+    vcf = VariantFile(chr3_vcf_path)
+    has_ref = "REF" in vcf.header.info["VRS_Allele_IDs"].description
+
+    for i, record in enumerate(vcf.fetch()):
+        print("~~~~~~~~~ row ", i, "~~~~~~~~~")
+
+        # use only alt VRS IDs, not REFs
+        vrs_allele_ids = record.info["VRS_Allele_IDs"]
+        print("vrs_allele_ids:", vrs_allele_ids)
+        if has_ref:
+            vrs_allele_ids = vrs_allele_ids[1:]
+
+        # for each alt ID, ensure stored allele counts match calculated allele counts
+        for alt_index, allele_id in enumerate(vrs_allele_ids):
+            print("alt id:", allele_id)
+            if not allele_id:
+                continue
+
+            caf = get_cohort_allele_frequency(
+                allele_id, chr3_vcf_path, phenotype_table=phenotype_table
+            )
+
+            print("alt_index", alt_index)
+            print("AC:", record.info["AC"][alt_index], caf["focusAlleleCount"])
+            print("AN:", record.info["AN"], caf["locusAlleleCount"])
+            ac = record.info["AC"][alt_index]
+            an = record.info["AN"]
+
+            assert (
+                caf["focusAlleleCount"] == ac
+            ), f"row {i} alt {alt_index} has different focus allele counts, expected {ac}, got {caf['focusAlleleCount']}"
+            assert (
+                caf["locusAlleleCount"] == an
+            ), f"row {i} alt {alt_index} has different locus allele counts, expected {an}, got {caf['locusAlleleCount']}"
+        if i == 5:
+            break
+
+
+def test_caf_generates_correct_allele_freq(
+    vrs_id_solo_alt, remote_chry_vcf_path, phenotype_table
+):
+    """test caf generation with default parameters and no phenotype specified"""
+    caf = get_cohort_allele_frequency(
+        vrs_id_solo_alt, remote_chry_vcf_path, phenotype_table=phenotype_table
+    )
+    print(json.dumps(caf))
+
+    # sanity checks
+    assert (
+        caf["type"] == "CohortAlleleFrequency"
+    ), f"object of type CohortAlleleFrequency not returned, returned {caf['type']} instead"
+    assert (
+        caf["focusAlleleCount"] <= caf["locusAlleleCount"]
+    ), f"Focus allele count ({caf['focusAlleleCount']}) is larger than locus allele count ({caf['locusAlleleCount']})"
+
+    print("focusAlleleCount:", caf["focusAlleleCount"])
+    print("locusAlleleCount:", caf["locusAlleleCount"])
+
+    # check allele frequency
+    expected_allele_freq = 0.0036
+    actual_allele_freq = round(caf["alleleFrequency"], 4)
+    assert (
+        actual_allele_freq == expected_allele_freq
+    ), f"incorrect allele frequency, expected {expected_allele_freq} got {actual_allele_freq}"
```

Comment:
fwiw there's a builtin pytest function to do this

https://docs.pytest.org/en/stable/reference/reference.html#pytest-approx